File: frontend/src/utils/timeSlot.ts
import { DAY_NAMES, DAY_START_HOUR, SLOT_COUNT, SLOT_MINUTES } from '@/constants'

export interface SlotOption {
  value: number
  label: string
}

/** 슬롯 번호를 "HH:MM"으로. SLOT_COUNT는 "24:00"이다. */
export function slotToTime(slot: number): string {
  if (slot < 0 || slot > SLOT_COUNT) {
    throw new RangeError(`슬롯 범위를 벗어났습니다: ${slot}`)
  }
  const minutes = DAY_START_HOUR * 60 + slot * SLOT_MINUTES
  const hh = String(Math.floor(minutes / 60)).padStart(2, '0')
  const mm = String(minutes % 60).padStart(2, '0')
  return `${hh}:${mm}`
}

/** "HH:MM"을 슬롯 번호로. 격자에 없는 시각은 거부한다. */
export function timeToSlot(text: string): number {
  const [hourText, minuteText] = text.split(':')
  const minutes = Number(hourText) * 60 + Number(minuteText)
  const offset = minutes - DAY_START_HOUR * 60
  if (offset < 0 || offset % SLOT_MINUTES !== 0) {
    throw new RangeError(`격자에 없는 시각입니다: ${text}`)
  }
  const slot = offset / SLOT_MINUTES
  if (slot > SLOT_COUNT) {
    throw new RangeError(`격자에 없는 시각입니다: ${text}`)
  }
  return slot
}

/** "월 09:00~13:00" — 목록과 툴팁에 쓰는 한 줄 설명. */
export function describeSchedule(day: number, start: number, end: number): string {
  return `${DAY_NAMES[day]} ${slotToTime(start)}~${slotToTime(end)}`
}

/** 시작·종료 드롭다운에 쓸 전체 선택지. */
export function slotOptions(): SlotOption[] {
  return Array.from({ length: SLOT_COUNT + 1 }, (_, slot) => ({
    value: slot,
    label: slotToTime(slot),
  }))
}

File: backend/app/timeslot.py
"""30분 단위 슬롯 정수와 사람이 읽는 시간 문자열 사이의 변환."""

from app.constants import DAY_NAMES, DAY_START_HOUR, SLOT_COUNT, SLOT_MINUTES


def slot_to_time(slot: int) -> str:
    """슬롯 번호를 "HH:MM"으로 바꾼다. slot=SLOT_COUNT는 "24:00"이다."""
    if not 0 <= slot <= SLOT_COUNT:
        raise ValueError(f"슬롯 범위를 벗어났습니다: {slot}")
    minutes = DAY_START_HOUR * 60 + slot * SLOT_MINUTES
    return f"{minutes // 60:02d}:{minutes % 60:02d}"


def time_to_slot(text: str) -> int:
    """"HH:MM"을 슬롯 번호로 바꾼다. 격자에 없는 시각은 거부한다."""
    hour_text, _, minute_text = text.partition(":")
    minutes = int(hour_text) * 60 + int(minute_text)
    offset = minutes - DAY_START_HOUR * 60
    if offset < 0 or offset % SLOT_MINUTES != 0:
        raise ValueError(f"격자에 없는 시각입니다: {text}")
    slot = offset // SLOT_MINUTES
    if slot > SLOT_COUNT:
        raise ValueError(f"격자에 없는 시각입니다: {text}")
    return slot


def describe(day_of_week: int, start_slot: int, end_slot: int) -> str:
    """"월 09:00~13:00" 형태의 한 줄 설명. 오류 메시지와 툴팁에 쓴다."""
    return (
        f"{DAY_NAMES[day_of_week]} "
        f"{slot_to_time(start_slot)}~{slot_to_time(end_slot)}"
    )


def overlaps(a_start: int, a_end: int, b_start: int, b_end: int) -> bool:
    """두 구간이 겹치는지 판정한다. 끝과 시작이 맞닿는 경우는 겹치지 않는다."""
    return a_start < b_end and b_start < a_end
